ytypes/util_types.go

```diff
@@ -134,6 +210,134 @@ func StringToType(t reflect.Type, s string) (reflect.Value, error) {
+	ykind := schema.Type.Kind
+	switch ykind {
+	// TODO(wenbli): case yang.Yempty: case yang.Ybits: case yang.Ybool:
+	case yang.Yint64, yang.Yint32, yang.Yint16, yang.Yint8:
+		bits, err := yangIntTypeBits(ykind)
+		if err != nil {
+			return reflect.ValueOf(nil), err
+		}
+		u, err := strconv.ParseInt(value, 10, bits)
+		if err != nil {
+			return reflect.ValueOf(nil), fmt.Errorf("unable to convert %q to %v", value, ykind)
+		}
+		// Although Convert can panic, we know that the type is an integer type and
+		// u must be a valid int type of the same length -- it is therefore impossible that
+		// Convert fails here.
+		return reflect.ValueOf(u).Convert(reflect.TypeOf(yangBuiltinTypeToGoType(ykind))), nil
+	case yang.Yuint64, yang.Yuint32, yang.Yuint16, yang.Yuint8:
+		bits, err := yangIntTypeBits(ykind)
+		if err != nil {
+			return reflect.ValueOf(nil), err
+		}
+		u, err := strconv.ParseUint(value, 10, bits)
+		if err != nil {
+			return reflect.ValueOf(nil), fmt.Errorf("unable to convert %q to %v", value, ykind)
+		}
+		// Although Convert can panic, we know that the type is an integer type and
+		// u must be a valid int type of the same length -- it is therefore impossible that
+		// Convert fails here.
+		return reflect.ValueOf(u).Convert(reflect.TypeOf(yangBuiltinTypeToGoType(ykind))), nil
+	case yang.Ybinary:
+		v, err := base64.StdEncoding.DecodeString(value)
+		if err != nil {
+			return reflect.ValueOf(nil), fmt.Errorf("error in DecodeString for \n%v\n for schema %s: %v", value, schema.Name, err)
+		}
+		return reflect.ValueOf([]byte(v)), nil
+	case yang.Ystring:
+		return reflect.ValueOf(value), nil
+	case yang.Ydecimal64:
+		floatV, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return reflect.ValueOf(nil), fmt.Errorf("unable to convert %q to %v: %v", value, ykind, err)
+		}
+		return reflect.ValueOf(floatV), nil
+	case yang.Yenum, yang.Yidentityref:
+		enumVal, err := enumStringToValue(parent, fieldName, value)
+		return reflect.ValueOf(enumVal), err
+	case yang.Yunion:
+		return stringToUnionType(schema, parent, fieldName, value)
+	}
+
+	return reflect.ValueOf(nil), fmt.Errorf("stringToKeyType: unsupported type %v for conversion from string %q, schema.Type: %v", ykind, value, schema.Type)
+}
+
+// stringToUnionType converts a string value into a suitable union type
+// determined by where it is located in the YANG tree.
+func stringToUnionType(schema *yang.Entry, parent interface{}, fieldName string, value string) (reflect.Value, error) {
+	util.DbgPrint("stringToUnionType value %v, into parent type %T field name %s, schema name %s", util.ValueStrDebug(value), parent, fieldName, schema.Name)
+	if !util.IsTypeStructPtr(reflect.TypeOf(parent)) {
+		return reflect.ValueOf(nil), fmt.Errorf("stringToKeyType: %T is not a struct ptr", parent)
+	}
+	parentT := reflect.TypeOf(parent)
+	dft, _ := parentT.Elem().FieldByName(fieldName)
```

Comment:
Do we need to handle the case here that this returns `false`? If I call this function with an invalid name then this seems like it might NPE at the next line?